### DISCUSSION TITLE
How to efficiently retrieve the cache status? `--dry-run` is very slow.

### DISCUSSION THREAD

**Discussion by thenbe** [vercel/turborepo#3875] [Help] [answered]
I want to know if a certain command is cached (aka turbo will not run it from scratch) before executing that command.

---

Example: I want to check if `site` is built.

I can use `--dry-run` to check, and that works. However it is incredibly slow (11 seconds in this particular repo):

```shell
pnpm turbo run build --filter=site --dry-run=json | jq '.tasks[-1].cacheState.local'
# true
```

At the same time, if I just do:

```shell
pnpm turbo run build --filter=site
```

turbo can almost _instantly_ tell if `build` is cached or if it needs to be re-executed. So obviously turbo doesn't actually need 11 seconds to get the info I want. `--dry-run` is probably doing much more stuff that I'm not interested in this particular instance.

So, how can I make turbo output the info I want without waiting that long?

> For context: If the answer is no, I want to abort and exit. There are some scripts (ex. testing) that hide the build output. So for those, I'm fine with them spinning up a dev server before running tests if they can grab it from cache and all is fresh, but I don't want them to actually build. In the latter case, I want to abort.

**Comment by mehulkar** [comment, marked as answer]
`--dry` is really the only option at the moment.

> turbo can almost instantly tell if build is cached or if it needs to be re-executed.

this isn't quite true. In order to find an existing cache, turbo has to calculate the hash of the current state of the inputs. If that's slow, then turbo can't know how to check whether cached outputs exist for that hash or not.

> However it is incredibly slow (11 seconds in this particular repo):

Could run the command you posted above and add the [`--cpuprofile=my-profile`](https://turbo.build/repo/docs/reference/command-line-reference#--cpuprofile) flag to see where it's spending so much time?
